### ISSUE TITLE
Informative error logging in case components can't be updated

### ISSUE THREAD

**Issue by immortaly007** [oxyno-zeta/prometheus-cachethq#6]
**Is your feature request related to a problem? Please describe.**
First of all, thanks for creating this code! It seems to be very versatile and exactly what I need! 
However, something is not going right for me, as the alerts don't arrive in cachet. I do see in the logs that the alertmanager notifications arrive (note that 172.29.0.8 is the IP for the alertmanager)

```
time="2020-11-08T23:32:46Z" level=debug msg="request started" client_ip=172.29.0.8 http_method=POST http_proto=HTTP/1.1 http_scheme=http remote_addr="172.29.0.8:58132" request_id=e3f6340c-4ef2-49e6-935c-c10e9e8b29cf uri="http://prometheus-cachet:8080/prometheus/webhook" user_agent=Alertmanager/0.20.0

time="2020-11-08T23:32:46Z" level=error msg="request complete" client_ip=172.29.0.8 http_method=POST http_proto=HTTP/1.1 http_scheme=http remote_addr="172.29.0.8:58132" request_id=e3f6340c-4ef2-49e6-935c-c10e9e8b29cf resp_bytes_length=31 resp_elapsed_ms=68.073926 resp_status=404 uri="http://prometheus-cachet:8080/prometheus/webhook" user_agent=Alertmanager/0.20.0
```
I don't know what is going wrong here. Is prometheus-cachethq not able to match the alert to a rule? Is it unable to connect to cachet? And in that case, is the problem the cachet URL or the API key? Or is it unable to match the component in the configuration to an existing component in cachet?

I think having more clear logging would greatly enhance the debugging experience. Unfortunately, I'm not a "Go" coder myself, so I'm unsure how I would go about debugging this, so I would be amazing if you could help!

**Describe the solution you'd like**
It would be nice if the logging would indicate in which part of the proces the error is originating, and, in more general terms, why my component in cachet isn't updating. 

**Describe alternatives you've considered**
I have tried to put the log level to "trace", but this didn't really reveal any additional information. 

**Additional context**
I'm running in docker. A snippet of the docker-compose file I use:

```yml
services:
  // Prometheus, node_exporter, cadvisor, alertmanager, etc here
  cachet:
    image: cachethq/docker:latest
    env_file:
      - ./cachet/config.env
    ports:
      - 8000
  prometheus-cachet:
    image: oxynozeta/prometheus-cachethq
    volumes:
      - ./cachet/prometheus-cachet:/conf
    ports:
      - 8080
      - 9090
```

And this is the config for prometheus-cachet I am using:
```yml
# Log configuration
log:
  # Log level
  level: trace
  # Log format
  format: text

# Server configurations
# server:
#   listenAddr: ""
#   port: 8080

# Cachet configuration
cachet:
  url: http://cachet:8000
  apiKey: <redacted>

# Targets
  - component:
      name: nginx-test
      group: WHEATLEY
      status: MAJOR_OUTAGE
    alerts:
      - name: blackbox_probe_failed
      - labels:
          instance: https://nginx-test.mydomain.com
```

And the labels for the (test) alert I'm triggering are:
```
alertname="blackbox_probe_failed"
instance="https://nginx-test.mydomain.com"
job="blackbox-http-2xx"
severity="page"
```

**Comment by oxyno-zeta** [comment]
Hello @immortaly007 ,

Thanks for your issue and happy to see that this code can help people :).

I must confess that I forgot to log the error before ending the request... That's pretty annoying, you are right.
I will make another version later today.

By the way, your configuration of prometheus-cachet seems weird to me. There isn't any `targets` key in your pasted one. Is it bad copy/paste or a real problem ?

Should be:
```yaml
# Log configuration
log:
  # Log level
  level: trace
  # Log format
  format: text

# Server configurations
# server:
#   listenAddr: ""
#   port: 8080

# Cachet configuration
cachet:
  url: http://cachet:8000
  apiKey: <redacted>

# Targets
targets:
  - component:
      name: nginx-test
      group: WHEATLEY
      status: MAJOR_OUTAGE
    alerts:
      - name: blackbox_probe_failed
      - labels:
          instance: https://nginx-test.mydomain.com
```

**Comment by immortaly007** [comment]
Thank you for responding so quickly!

Regarding the missing `targets:`, great eye for spotting that! In my real configuration there was actually another "target", but I guess when I took that one out of the configuration here, I deleted too much (it was a bit late last night when I opened this issue)

For your last comment, regarding the "`blackbox_probe_failed` OR contains labels", that is a mistake in my real configuration. To get it to trigger when "alert name is `blackbox_probe_failed` AND contains labels...", I should probably just remove the hyphen in front of the `labels` tag, like so, right?
```yml
# Targets
targets:
  - component:
      name: nginx-test
      group: WHEATLEY
      status: MAJOR_OUTAGE
    alerts:
      - name: blackbox_probe_failed
        labels:
          instance: https://nginx-test.example.com
```

Once again, thank you for responding so quickly, and I'm looking forward to the updated version!

**Comment by oxyno-zeta** [comment]
You are welcome :).

Yes your configuration seems better now. And you are correct for the "AND" case that you want. Maybe trying this new configuration can fix your problem.

I will work on a fix version for sure. I will keep in touch.

**Comment by immortaly007** [comment]
Thank you! I tried the new config but unfortunately, it gives the same problem
```
time="2020-11-09T08:57:46Z" level=debug msg="request started" client_ip=172.29.0.8 http_method=POST http_proto=HTTP/1.1 http_scheme=http remote_addr="172.29.0.8:49568" request_id=4f9135a2-c070-47f2-99a3-77b2563545c9 uri="http://prometheus-cachet:8080/prometheus/webhook" user_agent=Alertmanager/0.20.0


time="2020-11-09T08:57:46Z" level=error msg="request complete" client_ip=172.29.0.8 http_method=POST http_proto=HTTP/1.1 http_scheme=http remote_addr="172.29.0.8:49568" request_id=4f9135a2-c070-47f2-99a3-77b2563545c9 resp_bytes_length=31 resp_elapsed_ms=67.246344 resp_status=404 uri="http://prometheus-cachet:8080/prometheus/webhook" user_agent=Alertmanager/0.20.0
```

So unfortunately that wasn't the problem, and I'm still on the lookout for what is causing this, looking forward to any update you might release!

Small sidenote: In an earlier version of the config, I didn't specify the port of the cachet server in the config, and it logged `resp_status=500`. As it now logs a `resp_status=404`, something is different. Just to check: it is correct to specify the "root" of cachet as the `cachet.url`, and not e.g. the root of the api?

**Comment by oxyno-zeta** [comment]
Hum... I'm working on the log fix now, but, as it an old application (for me :) ), logs won't have all metadata that are present in others. This will require more time that I don't have for the moment. 

This application will need a rework to have more base features.

For your sidenote, check the example in this page: https://oxyno-zeta.github.io/prometheus-cachethq/docs/configuration.html . I think this will help you :)

**Comment by oxyno-zeta** [comment]
A new release has been done. You should see your error now in the log.

By the way, I found a way to have all metadata :) . This will be easier in case of multiple logs !

Can you test with the 1.1.1 ? Thanks !

**Comment by immortaly007** [comment]
Thank you so much for creating this update! I can now see that the error is "component not found". 
```
time="2020-11-09T18:07:16Z" level=error msg="component not found" client_ip=172.29.0.8 http_method=POST http_proto=HTTP/1.1 http_scheme=http remote_addr="172.29.0.8:55046" request_id=ba353a65-5d43-49b6-a6ca-68f405a99098 uri="http://prometheus-cachet:8080/prometheus/webhook" user_agent=Alertmanager/0.20.0
```

I checked out the code a bit, and this seems to indicate that it did manage to match my "alert" rule, but when querying cachet for the component, it couldn't find the component. I was able to confirm that it did do the query, from the cachet access logs:
 ```
172.29.0.9 - - [09/Nov/2020:18:22:16 +0000] "GET /api/v1/components?name=nginx-test&per_page=10000 HTTP/1.1" 200 277 "-" "Go-http-client/1.1" "-"
```
Interestingly, repeating this request from my Firefox I get:
```json
{
  "meta": {
    "pagination": {
      "total": 1,
      "count": 1,
      "per_page": 10000,
      "current_page": 1,
      "total_pages": 1,
      "links": {
        "next_page": null,
        "previous_page": null
      }
    }
  },
  "data": [
    {
      "id": 2,
      "name": "nginx-test",
      "description": "",
      "link": "https://nginx-test.basdado.com",
      "status": 1,
      "order": 0,
      "group_id": 1,
      "created_at": "2020-11-08 23:38:13",
      "updated_at": "2020-11-08 23:38:13",
      "deleted_at": null,
      "enabled": true,
      "meta": null,
      "status_name": "Operational",
      "tags": {
        "test": "test"
      }
    }
  ]
}
```
Access logs from my request:
```
172.29.0.10 - - [09/Nov/2020:18:36:23 +0000] "GET /api/v1/components?name=nginx-test&per_page=10000 HTTP/1.1" 200 277 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:82.0) Gecko/20100101 Firefox/82.0" "192.168.2.1"
```

My guess at this point is that maybe something changed in the cachet API since github.com/andygrunwald/cachet was last updated (I did look at the source code of that and didn't really see any changes tho...)?

TL;DR; The logging works, but I haven't figured out what is going wrong.

**Comment by immortaly007** [comment]
It was an issue with the group!

It turned out it was (again) a misread by me from the documentation. I specified the component group as `group: WHEATLEY`, but as can clearly be read in the documentation (https://oxyno-zeta.github.io/prometheus-cachethq/docs/configuration.html), it should be `groupName`.

A small thing that wasn't clear to me from said documentation tho, is that specifying the group is required IF the component is part of a group. I first tried without specifying the group, as the documentation said: 

> CachetHQ Component group name in which the component is. This is useful when you have multiple components with the same name into multiple groups.

Since I didn't have multiple components with the same name, I figured I didn't need this specification at first. However, looking at the code (in pkg\cachethq\private.go), I see that it actually filters on this group:
```go
// Filter components by groups
	// Client doesn't manage group id equal to 0 for no groups...
	for _, comp := range c.Components {
		if comp.GroupID == grpID {
			return &comp, nil
		}
	}
```

So a final small suggestion would be to update the documentation for this field to something like:

 > CachetHQ Component group name in which the component is. Only required if your component is part of a group, and can be useful when you have multiple components with the same name into multiple groups.

Alternatively, the code could be updated. I don't really know Go, but something like this on line 58 of `pkg\cachethq\private.go` might work:
```go
// If no group filter is specified
if groupName == "" {
    if len(c.Components) > 1 {
         // Log: multiple components matching the filter. Consider specifying a group.
   } 
    return &c.Components[0], nil;
}
```

I hope you don't think I'm overstepping by making code suggestions without knowing Go. 

Anyway, thank you so much for your help! This is a great open source tool for people like me who want to publish prometheus alerts directly to a public status page!

**Comment by oxyno-zeta** [comment]
When a group is specified, the group id is searched based on the name and the id is used after to find component with it.

Can you try by removing your group first ? Just to test that prometheus-cachet can talk to cachet correctly. If this works, the problem comes from your group.

**Comment by oxyno-zeta** [comment]
No problem @immortaly007 . Thanks for using this tool. I published it for people to use it !

I will think about the go code you proposed. 

For the configuration documentation, you are correct. I will fix it. Thanks !

Can I close this issue ?

**Comment by immortaly007** [comment]
Thank you! Yes, you can close the issue.

After I'm done configuring it for myself, I will also create a public repo with my full docker-compose setup (prometheus, node-exporter, cachet, and this tool), and a bit of documentation, so that hopefully more people will be able to see and use this tool "in action" and create similar setups that include this tool!

**Comment by oxyno-zeta** [comment]
Cool ! Maybe you can put your project link in this issue.
